src/System.Net.Http/tests/UnitTests/HPack/HPackRoundtripTests.cs

```diff
@@ -0,0 +1,180 @@
+                }
+            }
+
+            return buffer.ActiveMemory;
+
+            void WriteBytes(ReadOnlySpan<byte> bytes)
+            {
+                if (bytes.Length > buffer.AvailableLength)
+                {
+                    buffer.EnsureAvailableSpace(bytes.Length);
+                    FillAvailableSpaceWithOnes(buffer);
+                }
+
+                bytes.CopyTo(buffer.AvailableSpan);
+                buffer.Commit(bytes.Length);
+            }
+
+            void WriteLiteralHeaderValues(string[] values, string separator)
+            {
+                int bytesWritten;
+                while (!HPackEncoder.EncodeStringLiterals(values, separator, buffer.AvailableSpan, out bytesWritten))
+                {
+                    buffer.EnsureAvailableSpace(buffer.AvailableLength + 1);
+                    FillAvailableSpaceWithOnes(buffer);
+                }
+
+                buffer.Commit(bytesWritten);
+            }
+
+            void WriteLiteralHeader(string name, string[] values)
+            {
+                int bytesWritten;
+                while (!HPackEncoder.EncodeLiteralHeaderFieldWithoutIndexingNewName(name, values, HttpHeaderParser.DefaultSeparator, buffer.AvailableSpan, out bytesWritten))
+                {
+                    buffer.EnsureAvailableSpace(buffer.AvailableLength + 1);
+                    FillAvailableSpaceWithOnes(buffer);
+                }
+
+                buffer.Commit(bytesWritten);
+            }
+
+            // force issues related to buffer not being zeroed out
+            void FillAvailableSpaceWithOnes(ArrayBuffer buffer) => buffer.AvailableSpan.Fill(0xff);
+        }
+
+        // adapted from header deserialization code in Http2Connection.cs
+        private static HttpHeaders HPackDecode(Memory<byte> memory)
+        {
+            var header = new HttpRequestHeaders();
+            var hpackDecoder = new HPackDecoder(maxDynamicTableSize: 0, maxResponseHeadersLength: HttpHandlerDefaults.DefaultMaxResponseHeadersLength * 1024);
+
+            hpackDecoder.Decode(memory.Span, true, ((_, name, value) => HeaderHandler(name, value)), null);
+
+            return header;
+
+            void HeaderHandler(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
+            {
+                if (!HeaderDescriptor.TryGet(name, out HeaderDescriptor descriptor))
+                {
+                    throw new HttpRequestException(SR.Format(SR.net_http_invalid_response_header_name, Encoding.ASCII.GetString(name)));
+                }
+
+                string headerValue = descriptor.GetHeaderValue(value);
+
+                header.TryAddWithoutValidation(descriptor, headerValue.Split(',').Select(x => x.Trim()));
+            }
+        }
+    }
+
+    public static class AssertExtensions
+    {
+        public static void AssertForall<T>(this IEnumerable<T> collection, Action<T> body) => Assert.All(collection, body);
```

Comment:
Is this worth it just to enable a LINQ statement?

Reply:
I'd rather minimize the complexity and just have the test use the xunit methods directly. I personally don't that extension methods inhibit rather than help readability.

Reply:
It'll still be mixed in with LINQ, however I went for the extension because the xunit way to express it produces a rather abominable result:
```
Assert.All(expected.Value.Zip(actual.Value), x => Assert.Equal(x.First, x.Second));
```
Nevertheless, I switched back into that.

Reply:
> Nevertheless, I switched back into that.

Thanks.